File: tests/fixtures/ts-sample/src/features/todos/controller.ts
// VIOLATION: features/ must not import from data/sources/.
import { TodoStore } from '../../data/sources/todoStore';

// VIOLATION: features/ must not import from data/repositories/ either.
import { TodoRepository } from '../../data/repositories/todoRepository';

// OK: features importing other features is allowed by these rules.
import { OtherThing } from '../shared/util';

export class TodosController {
  constructor(private store: TodoStore, private repo: TodoRepository) {}
}

File: tests/fixtures/py-sample/src/myapp/features/todos/controller.py
# VIOLATION: features/ may not import from data/sources/.
from src.myapp.data.sources.todo_store import TodoStore

# VIOLATION: same rule applies to data/repositories/.
import src.myapp.data.repositories.todo_repository

# OK: sibling import.
from .helpers import normalize


def list_todos():
    return TodoStore().list()
